### PR TITLE
Extract SchemaResolver from MarshmallowPlugin

### PR DESCRIPTION
**Pull request by Bangertm**
Extracts logic for resolving schemas in OpenAPI objects and enhannces documentation.

Might have gone a little bit overboard on the documentation, but I thought it would helpful to provide concrete examples of how to construct dictionaries for conversion.

### CODE REVIEW

File: src/apispec/ext/marshmallow/schema_resolver.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
+from .common import resolve_schema_instance
+
+
+class SchemaResolver:
+    """Resolve Marshmallow Schemas in OpenAPI components and translate to OpenAPI
+    `schema objects
+    <https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schema-object>`_,
```
Comment (sloria) [inline src/apispec/ext/marshmallow/schema_resolver.py]: I think we've been linking to 3.0.2 in the rest of the docs.

```suggestion
    <https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schema-object>`_,
```

File: src/apispec/ext/marshmallow/schema_resolver.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
+from .common import resolve_schema_instance
+
+
+class SchemaResolver:
+    """Resolve Marshmallow Schemas in OpenAPI components and translate to OpenAPI
```
Comment (sloria) [inline src/apispec/ext/marshmallow/schema_resolver.py]: Lowercase marshmallow

```suggestion
    """Resolve marshmallow Schemas in OpenAPI components and translate to OpenAPI
```

File: src/apispec/ext/marshmallow/schema_resolver.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        self.openapi_version = openapi_version
+        self.converter = converter
+
+    def resolve_parameters(self, parameters):
+        """Resolve Marshmallow Schemas in a list of OpenAPI `Parameter Objects
+        <https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#parameter-object>`_.
+        Each parameter object that contains a Schema will be translated into
+        one or more Parameter Objects.
+
+        If the value of a `schema` key is Marshmallow Schema class, instance or
```
Comment (sloria) [inline src/apispec/ext/marshmallow/schema_resolver.py]:
```suggestion
        If the value of a `schema` key is a marshmallow Schema class, instance or
```

**Review by sloria — COMMENTED** [review]
Just drive-by nits. Will probably leave this up to @lafrech to give the ✅

File: src/apispec/ext/marshmallow/schema_resolver.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        a string that resolves to a Schema Class each field in the Schema will
+        be expanded as a seperate Parameter Object.
```
Comment (lafrech) [inline src/apispec/ext/marshmallow/schema_resolver.py]: s/seperate/separate

**Comment by lafrech** [comment]
I don't really see a benefit to this change. It's does not help customizing (subclassing), for instance, like the other alike changes we introduced recently do.

But I see no objection either.

We can merge this, if only for the added documentation.

Would this be the last change before we release v3.0?

**Comment by lafrech** [comment]
Nits fixed.

So, @sloria @Bangertm, we merge this and #501 and we ship a 3.0?

**Comment by sloria** [comment]
Sounds good to me

**Comment by lafrech** [comment]
Alright. Merged.

I shall release 3.0 this WE or beginning of next week.

Do we need to add a changelog entry for all those refactors (`FieldConverterMixin`, `SchemaResolver`, cleaner functions becoming `APISpec` methods)? Could some of them be considered breaking changes?

**Comment by lafrech** [comment]
@sloria, @Bangertm, I just updated the CHANGELOG and pushed 3.0.0!

**Comment by sloria** [comment]
Awesome! Thank you

**Pull request merged**